sdk/src/logs/logger.cc

```diff
@@ -34,25 +50,70 @@ void Logger::log(const opentelemetry::logs::LogRecord &record) noexcept
     return;
   }
 
-  // TODO: Sampler logic (should include check for minSeverity)
+  // TODO: Sampler (should include check for minSeverity)
+
+  auto recordable = processor->MakeRecordable();
+  if (recordable == nullptr)
+  {
+    // TODO: Error diagnostics should indicate "recordable creation failed" to user
+    return;
+  }
+
+  // Populate recordable fields
+  recordable->SetTimestamp(timestamp);
+  recordable->SetSeverity(severity);
+  recordable->SetName(name);
+  recordable->SetBody(body);
 
-  /**
-   * Convert the LogRecord to the heap first before sending to processor.
-   * TODO: Change the API log(LogRecord) function to log(*LogRecord) so the following line
-   * converting record a heap variable can be removed
-   */
-  auto record_pointer =
-      std::unique_ptr<opentelemetry::logs::LogRecord>(new opentelemetry::logs::LogRecord(record));
+  resource.ForEachKeyValue([&](nostd::string_view key,
+                               opentelemetry::common::AttributeValue value) noexcept {
+    recordable->SetResource(key, value);
+    return true;
+  });
 
-  // TODO: Do not want to overwrite user-set timestamp if there already is one -
-  // add a flag in the API to check if timestamp is set by user already before setting timestamp
+  attributes.ForEachKeyValue([&](nostd::string_view key,
+                                 opentelemetry::common::AttributeValue value) noexcept {
+    recordable->SetAttribute(key, value);
+    return true;
+  });
 
-  // Inject timestamp if none is set
-  record_pointer->timestamp = core::SystemTimestamp(std::chrono::system_clock::now());
-  // TODO: inject traceid/spanid later
+  // Inject trace_id/span_id/trace_flags if none is set by user
+  auto provider     = opentelemetry::trace::Provider::GetTracerProvider();
+  auto tracer       = provider->GetTracer(logger_name_);
+  auto span_context = tracer->GetCurrentSpan()->GetContext();
+
+  // Traceid
+  if (!trace_id.IsValid() && span_context.trace_id().IsValid())
+  {
+    recordable->SetTraceId(span_context.trace_id());
+  }
+  else
+  {
+    recordable->SetTraceId(trace_id);
```

Comment:
with the new check above, it's possible you're setting trace_id here if `span_context.trace_id().Isvalid()` returns false, i.e. it could be invalid.

Might be easier just do:

```
if trace_id valid => setTraceId
else if context trace id valid => setTraceId
else // nothing.
```

Reply:
This is much cleaner :) Thanks, updated as above (minus the empty `else` block)